### PR TITLE
Prevent loading a world with unresolved dependencies

### LINKED ISSUE #12217: Prevent loading a world with unresolved dependencies
## Problem

Currently, Minetest will disable mods that have unfulfilled dependencies and then still load the game. This has two problems:

* It's a bad user-experience as players can end up playing a game without the mods they enabled
* This causes problems when a mod opt depends on another mod with unfulfilled dependencies, see #8601 
* It results in a complicated mod dependency algorithm

## Solutions

Mods being implicitly disabled when they have unfulfilled dependencies/cycles should be deprecated in 5.6.0.

In 5.7.0, we should throw a hard error when a mod doesn't have a fulfilled dependency, or if there's cycles. This will allow the algorithm to be greatly simplified. In the case given in the issue, Minetest will show an error and return to the menu - the user would need to disable mod E. The Config Mods menu should also show errors such as cycles by highlighting mods red.

The hard error is partially implemented by #8059

image

## Additional context

Supersedes #8601, #11438, and #10486

Related #9977

### PR DESCRIPTION
**Pull request by rubenwardy**
Fixes #12217

## To do

This PR is a Ready for Review.

## How to test

1. Enable some mods but not their dependencies
2. Attempt to load the world

### CODE REVIEW

File: src/content/mod_configuration.cpp [hunk 1]
```diff
@@ -21,16 +21,26 @@ with this program; if not, write to the Free Software Foundation, Inc.,
...
+
+std::string ModConfiguration::getUnsatisfiedModsError() const
 {
+	std::ostringstream error;
+	error << gettext("Some mods have unsatisfied dependencies:") << std::endl;
+
 	for (const ModSpec &mod : m_unsatisfied_mods) {
-		errorstream << "mod \"" << mod.name
-					<< "\" has unsatisfied dependencies: ";
+		error << " - " << fmtgettext("%s is missing:", mod.name.c_str());
```
Comment (sfan5) [inline src/content/mod_configuration.cpp:33]: needs an annotation or nobody will ever translate this correctly

Comment (rubenwardy) [reply src/content/mod_configuration.cpp:33]: How do I do annotations?

Comment (sfan5) [reply src/content/mod_configuration.cpp:33]: `//~ text here`

Comment (Zughy) [reply src/content/mod_configuration.cpp:33]: @rubenwardy :eyes:

**Review by sfan5 — APPROVED** [review]
LGTM

**Pull request merged**